### PR TITLE
Radiation rework and subsystem

### PR DESCRIPTION
**Pull request by ninjanomnom**
:cl: ninjanomnom
refactor: Radiation has been completely overhauled.
add: A new radiation subsystem and spreading mechanics.
add: Walls and other dense objects insulate you from radiation.
add: Geiger counters now store the last burst of radiation so you can view it at your leisure or show it to someone. Examine it.
add: Geiger counters can check mobs for contaminated objects. Scan yourself before you leave to make sure you aren't carrying dangerous radioactive items.
soundadd: Geiger counters have realistic sounds and the radiation pulse spam in chat has been replaced.
balance: Radiation is more deadly and causes burns at high intensities, WEAR YOUR RADSUITS.
balance: However residue radiation is far slower acting and kills you with toxin.
balance: Engineering holosigns have a light amount of radiation insulation.
balance: Rad collectors are nerfed. No more supercharging with pressurized plasma. No more collector spam either.
balance: Engine output is a lot more stable as a result of collector changes.
balance: Monkeys need more rads and take more time to turn into gorillas.
tweak: Over 100% on the DNA computer means your subject is now taking damage.
/:cl:

The goal of this is to remove snowflake applications of radiation and make it act realistically. With this system radiation can be insulated and it also now obeys the inverse square law. Overall this is a buff to radiation.

TODO:
- [x] Write the basic systems
- [x] Add radiation insulation to various turfs and structures
- [x] Find and convert all snowflake usages of radiation damage
- [x] Finish up radiation contamination
- [x] Reduce resource usage (Might need to do more)
- [x] Fix runtimes
- [x] Remap supermatter chambers (Will be done when the code is approved)
- [x] Test and balance

Issues:
- [x] ~~The SM gas interactions don't work well with power scaling~~ Fixed with some tweaks
- [x] ~~Contaminated tiles are building up rads when they shouldn't be~~ Bad use of log()
- [x] ~~Figure out exactly why the range gets so high (more insulation maybe?)~~ See above
- [x] ~~The high end is too high and the low end too low for mobs to handle well~~ Rewrote mob rad handling
- [x] ~~Genetics radiation needs to be fixed~~ Rad damage is matched with the others
- [x] ~~Mobs aren't getting toxin damage~~ Bad return

### CODE REVIEW

**Comment by DoctoreMad** [comment]
Separation of alpha, beta and gamma radiation when?

**Comment by ninjanomnom** [comment]
Out of scope. This doesn't change what radiation does, just how it acts.

**Comment by Jordie0608** [comment]
Half-life memes when?

**Comment by ninjanomnom** [comment]
I could extremely easily make radioactive tainting with a half-life :^)

File: code/game/objects/structures/mineral_doors.dm
Comment (ChangelingRain) [inline code/game/objects/structures/mineral_doors.dm]: make the insulation level a var and only create the component if there's insulation

File: code/game/atoms.dm
Comment (optimumtact) [inline code/game/atoms.dm]: this is retarded, just call SendSignal on the atom

File: code/datums/components/rad_insulation.dm
Comment (Cyberboss) [inline code/datums/components/rad_insulation.dm]: set the default as the proc var default instead of on the datum

File: code/datums/radiation_wave.dm
Comment (Cyberboss) [inline code/datums/radiation_wave.dm]: Use place.GetAllContents

File: code/game/objects/structures/mineral_doors.dm
Comment (optimumtact) [reply code/game/objects/structures/mineral_doors.dm]: that's just needless var bloat

**Review by Cyberboss — CHANGES_REQUESTED** [review]

File: code/game/objects/structures/mineral_doors.dm
Comment (Cyberboss) [inline code/game/objects/structures/mineral_doors.dm]: wat

Comment (Cyberboss) [reply code/game/objects/structures/mineral_doors.dm]: sir, you know qdel exists right?

**Review by Cyberboss — CHANGES_REQUESTED** [review]

File: code/game/atoms.dm
Comment (ninjanomnom) [reply code/game/atoms.dm]: This proc already exists and is used down the inheritance chain

File: code/datums/radiation_wave.dm
Comment (Cyberboss) [inline code/datums/radiation_wave.dm]: use START_PROCESSING

File: code/game/objects/structures/mineral_doors.dm
Comment (ninjanomnom) [reply code/game/objects/structures/mineral_doors.dm]: ...right, yes. I was rushing this part and was going off of component documentation looking for something like RemoveComponent() and it said destroy was public so I used it >.>

File: code/game/atoms.dm
Comment (Cyberboss) [reply code/game/atoms.dm]: since you're rewriting radiation you can get rid of the proc

**Review by Cyberboss — DISMISSED** [review]

File: code/game/atoms.dm
Comment (ninjanomnom) [reply code/game/atoms.dm]: Alright, time for a bunch of new components then

**Comment by banthebantz** [comment]
Fucking finally, this is very good

File: code/datums/components/radioactive.dm
Comment (Qustinnus) [inline code/datums/components/radioactive.dm]: wew lad

**Comment by KorPhaeron** [comment]
Stale, tell me when it's done @ninjanomnom I was looking forward to this

**Comment by ninjanomnom** [comment]
Repeating here what I said to kor in irc:

I'd prefer to not testmerge this sort of thing until the code has been mostly approved of by a maintainer because of the impact this could have on a round. And I'd also prefer to not do map changes until that happens. If I should go about this differently, let me know.

**Comment by optimumtact** [comment]
can't review it without a test merge

**Comment by ninjanomnom** [comment]
Fair enough I suppose, I'll work on the map changes tomorrow when I'm not on a caffeine life support at work.

Let's fuck up some poor engineer's day

File: code/game/atoms.dm
Comment (Cyberboss) [reply code/game/atoms.dm]: HA RADtarded i just got that

**Comment by qbmax32** [comment]
Live news from Bagil: It's randomly killing people and at least 15 people have dropped dead without warning. Fix plz XD

**Comment by Jazaen** [comment]
Not only 15 people, but almost entire station. And after admins resurrect them, they walk a few steps and die again.

**Comment by qbmax32** [comment]
This just in: geigar counters recieving 124005 rads and everyone gets ded including the AI and borgs, Its outta control

**Comment by Qustinnus** [comment]
guys protip: stand next to the SM and you'll be the safest. The further you're away, the deadlier it is. :^)

**Comment by qbmax32** [comment]
ill donate to FTL if you save us from this hell

**Comment by Qustinnus** [comment]
alright it's fixed in the next update. Don't forget to donate.

**Comment by ninjanomnom** [comment]
Get out of my pr floyd

**Comment by Qustinnus** [comment]
never, I live here now.

**Comment by Supermichael777** [comment]
Literally instant death due to infectious radiation/10

**Comment by ninjanomnom** [comment]
changes for next testmerge update:

fixes the geiger counter
buffs collector power gen
nerfs sm radiation
slightly buffs rad insulation
nerfs rad damage (and its burn)
raised the minimum radiation

**Comment by WJohn** [comment]
>nuke ops land at the very north of the station as far as possible from the SM
>one of the ops INSTANTLY explodes from rads
>OH FUCK ABORT ABORT
>rest of the ops die seconds after returning to their base

good times

**Comment by davethwave** [comment]
I wish I could have recorded that round, it was the only round that made me laugh so hard before.

**Comment by GuyonBroadway** [comment]
I wanted the great uranium wall holocast to come back but this has gone from Holocaust to Imperial galactic ethnic cleansing program.

**Comment by Mey-Ha-Zah** [comment]
I was a miner the shift the rad happened. I heard nobody talking over radio and went back to deposit my ores, wandered into the halls and saw three bodies and a red Geiger counter. My brain clicked and told me to run. I got about 8 tiles before the rad surge hit me and a died from like a million burn damage and was instantly husked.

10/10

**Comment by Cyberboss** [comment]
How we doing here?

**Comment by ninjanomnom** [comment]
It's been testmerged for a while which has been helpful for finding all the bugs and rad sources I missed. I still need to make showers clean contamination and fix genetic rad values. Then it's done. All the balancing is complete (At least enough for this pr)

I've got some planned stuff for radiation after this but that'll wait, this is a big enough pr as it is

**Comment by ninjanomnom** [comment]
On second thought I'm leaving shower changes to the followup pr. Contamination goes away after a while anyway and showers are a mess of a code that need a rewrite; too big for this pr.

File: code/__HELPERS/radiation.dm
Comment (Cyberboss) [inline code/__HELPERS/radiation.dm]: can you make this iterative instead of recursive please

File: code/__HELPERS/radiation.dm
Comment (Cyberboss) [inline code/__HELPERS/radiation.dm]: log = FALSE

File: code/controllers/subsystem/radiation.dm
Comment (Cyberboss) [inline code/controllers/subsystem/radiation.dm]: Use PROCESSING_SUBSYSTEM_DEF

File: code/controllers/subsystem/radiation.dm
Comment (Cyberboss) [inline code/controllers/subsystem/radiation.dm]: Then this entire proc isn't needed

File: code/datums/components/rad_insulation.dm
Comment (Cyberboss) [inline code/datums/components/rad_insulation.dm]: Initialize on components is an abstract proc so this isn't required

File: code/datums/components/radioactive.dm
Comment (Cyberboss) [inline code/datums/components/radioactive.dm]: ditto

File: code/datums/components/radioactive.dm
Comment (Cyberboss) [inline code/datums/components/radioactive.dm]: CRASH instead of warning. I still haven't handled components deleting themselves in Initialize(). Not sure if I should

File: code/datums/components/radioactive.dm
Comment (Cyberboss) [inline code/datums/components/radioactive.dm]: Not possible

File: code/datums/components/radioactive.dm
Comment (Cyberboss) [inline code/datums/components/radioactive.dm]: ..() undeeded

File: code/datums/components/radioactive.dm
Comment (Cyberboss) [inline code/datums/components/radioactive.dm]: ..() undeeded

File: code/datums/components/radioactive.dm
Comment (Cyberboss) [inline code/datums/components/radioactive.dm]: ..() undeeded

File: code/datums/components/radioactive.dm
Comment (Cyberboss) [inline code/datums/components/radioactive.dm]: var/atom/P = parent to avoid datum var lookups

File: code/datums/components/radioactive.dm
Comment (Cyberboss) [inline code/datums/components/radioactive.dm]: replace these with it too

File: code/datums/components/radioactive.dm
Comment (Cyberboss) [inline code/datums/components/radioactive.dm]: var/list/out = list()

File: code/datums/components/radioactive.dm
Comment (Cyberboss) [inline code/datums/components/radioactive.dm]: out.Join()

File: code/datums/radiation_wave.dm
Comment (Cyberboss) [inline code/datums/radiation_wave.dm]: This retval isn't useful

File: code/datums/radiation_wave.dm
Comment (Cyberboss) [inline code/datums/radiation_wave.dm]: cache move_dir instead of reading it twice

File: code/datums/radiation_wave.dm
Comment (Cyberboss) [inline code/datums/radiation_wave.dm]: ditto for master_turf

File: code/datums/radiation_wave.dm
Comment (Cyberboss) [inline code/datums/radiation_wave.dm]: see above

File: code/datums/radiation_wave.dm
Comment (Cyberboss) [inline code/datums/radiation_wave.dm]: I feel like this should be a GLOB typecache because I swear it's used more than just here

Comment (Cyberboss) [reply code/datums/radiation_wave.dm]: Also, just typecacheof(/turf)

File: code/game/atoms.dm
Comment (Cyberboss) [inline code/game/atoms.dm]: Can drop the return as it's not checked

File: code/game/objects/structures/mineral_doors.dm
Comment (Cyberboss) [inline code/game/objects/structures/mineral_doors.dm]: TRUE

File: code/game/objects/structures/mineral_doors.dm
Comment (Cyberboss) [inline code/game/objects/structures/mineral_doors.dm]: Instead of deleting new components like this, Add a common proc that you just override for /uranium

File: code/game/objects/structures/mineral_doors.dm
Comment (Cyberboss) [inline code/game/objects/structures/mineral_doors.dm]: ditto

File: code/modules/power/supermatter/supermatter.dm
Comment (Cyberboss) [inline code/modules/power/supermatter/supermatter.dm]: no need for named arg log

File: code/modules/recycling/disposal-unit.dm
Comment (Cyberboss) [inline code/modules/recycling/disposal-unit.dm]: The way this works is it asks SSatoms to call atom/LateInitialize() after it has Initialized all atoms. You didn't define one

**Review by Cyberboss — CHANGES_REQUESTED** [review]
I like it

File: code/datums/radiation_wave.dm
Comment (ninjanomnom) [reply code/datums/radiation_wave.dm]: I'd prefer to leave this here because I'm pretty sure I'll need to add more entries

File: code/modules/recycling/disposal-unit.dm
Comment (ninjanomnom) [reply code/modules/recycling/disposal-unit.dm]: No it exists, I just moved this line from a duplicate Initialize

File: code/__HELPERS/radiation.dm
Comment (ninjanomnom) [reply code/__HELPERS/radiation.dm]: The recursive method has been proven faster in #31027

File: code/game/atoms.dm
Comment (Cyberboss) [inline code/game/atoms.dm]: trailing return

**Review by Cyberboss — CHANGES_REQUESTED** [review]

**Review by Cyberboss — DISMISSED** [review]

**Review by Cyberboss — APPROVED** [review]

**Comment by GuyonBroadway** [comment]
Pretty rad.

**Comment by davethwave** [comment]
congratulations

**Comment by evsey9** [comment]
REVERT THIS SHIT RIGHT FUCKING NOW
ITS BROKEN

**Comment by Hathkar** [comment]
Currently extremely broken. A single pulse from genetics sends the rad levels to around 500%. Once you get any in your system, it's pretty much impossible to remove any. Revert or fix.

**Comment by praisenarsie** [comment]
...this was merged 3 days ago, guys.

**Comment by evsey9** [comment]
Also, this breaks the singularity
Not enough power for the station.

**Comment by davethwave** [comment]
perhaps make the singulo emit more rads or make rads spread with less loss in space to fix this?

**Comment by ike709** [comment]
There is already an open PR to fix genetics.

**Comment by ike709** [comment]
And issues open for the singularity.

**Comment by msgerbs** [comment]
This is to this day the worst PR to ever grace TG.

**Comment by optimumtact** [comment]
did you die to radiation?

**Comment by GuyonBroadway** [comment]
I dunno, it's pretty rad.

**Comment by msgerbs** [comment]
Damn right I did, except I didn't just die, I spent 20 minutes trying not to die, all because I tried to arrest somebody who had been in engineering. Making radiation some kind of contagious disease was and still is a horrible idea.

**Comment by optimumtact** [comment]
he died :joy:

**Comment by FancyCyclops** [comment]
This is how the fall began. How we went down the path of failure; the mark of the beginning of the end.

This PR and its subsequent merging was a mistake; it marks the beginning of a new era; an era of failure and headcoder griefing.

One can only pray for its excision from the code.

**Pull request merged**
